lambdify: Included scipy backend automatically and added relevant functions

ramvenkat98 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests .-->

Fixes #7229 
Fixes #14973 
Fixes #14975 
Referred to PR 10319

#### Brief description of what is fixed or changed

Added relevant scipy.special functions to sympy/sympy/printing/pycode.py
Added use of SciPy printer to lambdify
Added fix so that lambdify automatically uses scipy if available
Added tests to sympy/utilities/tests/test_lambdify.py
Fixed error in sympy/sympy/printing/pycode.py, replacing lgamma with
loggamma in the _not_in_numpy list

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. If there is no release notes entry for this PR,
write "NO ENTRY". The bot will check your release notes automatically to see
if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* utilities
  * lambdify uses scipy automatically if available
<!-- END RELEASE NOTES -->

ramvenkat98 committed:
lambdify: Included scipy automatically and added functions

Added relevant scipy.special functions to sympy/sympy/printing/pycode.py
Added use of SciPy printer to lambdify
Added fix so that lambdify automatically uses scipy if available (issue 7229)
Added tests to sympy/utilities/tests/test_lambdify.py
Fixed error in sympy/sympy/printing/pycode.py, replacing lgamma with
loggamma in the _not_in_numpy list

ramvenkat98 committed:
Modified earlier commit to solve with mpmath correctly

Modified earlier commit so that mpmath would not be affected

bjodah reviewed (commented):
Looks good to me. Perhaps we should set numpy.random.seed

asmeurer commented on sympy/utilities/lambdify.py:
```diff
@@ -360,14 +364,19 @@ def lambdify(args, expr, modules=None, printer=None, use_imps=True,
+                modules = ["numpy"]
         else:
-            modules = ["numpy"]
+            modules = ["scipy"]
```
Shouldn't it be `["scipy", "numpy"]`?

asmeurer commented:
You need to write out "fixes" separately for each issue in the PR description so they will all be automatically closed.

ramvenkat98 committed:
Added numpy to modules, Changed if statements, added numpy random seed

Added numpy to namespace with scipy
Changed printer order
Added random seed in lambdify test

ramvenkat98 commented:
@asmeurer @bjodah 

I've updated the pull request according to the review:
1) Set numpy.random.seed(0) before the random tests
2) Also adding "numpy" to the modules list in lambdify when scipy is present (and changing the order of Printer selection accordingly to look for SciPy printer before NumPy printer)
3) Wrote out fixes separately for each issue in the PR

However, exactly one Travis test is failing right now and it seems to be a random failure unrelated to the fix- could this test be re-run?

Thanks

bjodah commented:
Yes, I restarted the test, thank you for telling.

ramvenkat98 commented:
@asmeurer @bjodah 

Hi,

are these fixes ok and can the pull request be merged?

Thanks again

ramvenkat98 commented:
@asmeurer @bjodah 

Hi,

may I have an update on the status of this pull request? Should I fix anything more or is it ok to be merged? (Sorry for commenting multiple times in a row, I figured it's been a few days since the last activity so I brought it up again).

Thanks

bjodah commented:
@ramvenkat98 thank you very much for your contribution (and sorry for the wait)

GProtoZeroW commented:
@ramvenkat98 
I just tried a simple call on of the Bessel function of the first kind in sympy and tried to lambdify it to scipy's special.jv and it did not work. Is there something that I am missing to invoke the translation from sympy to scipy in the improved lambdfy in sympy1.3

```
from sympy import *
import scipy.special as SciSpec

#scipy ref 
#https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.jv.html#scipy.special.jv
Order=1; Arg=1.2
SciSpec.jv(Order, Arg)
>>>0.49828905756721553

#sympy call
nu, z=symbols('nu, z')
#https://docs.sympy.org/latest/modules/functions/special.html?highlight=bessel#module-
sympy.functions.special.bessel
b=besselj(nu, z); b
N(b.subs({nu:Order, z:Arg}))
>>>0.498289057567215

bN=lambdify((nu, z), b, dummify=False)
bN(Order, Arg)
>>>NameError                                 Traceback (most recent call last)
<ipython-input-10-3f7a3bb1d5f0> in <module>()
      1 bN=lambdify((nu, z), b, dummify=False)
----> 2 bN(Order, Arg)

~/anaconda3/lib/python3.6/site-packages/numpy/__init__.py in <lambda>(nu, z)

NameError: name 'besselj' is not defined
```

ramvenkat98 commented:
@GProtoZeroW @asmeurer @bjodah 

When using the version of sympy pulled from Github, I don't get this error; however, when using sympy installed as the Anaconda package, I do. Perhaps, this is because the issue has already been fixed on Github but did not make it to the currently released version?

moorepants commented:
SymPy 1.1.1 is available from the default channel for Anaconda. If you want SymPy >=1.2 you need to install from the conda-forge channel (at least until the default get updated).

GProtoZeroW commented:
@ramvenkat98 I did the Anaconda install I will have to do the github one to test out. And thanks for adding scipy to lambdfiy it's been needed. Just two questions: one does `Heaviside` from sympy now translate over to numpy's  `heaviside` (https://docs.scipy.org/doc/numpy-1.14.0/reference/generated/numpy.heaviside.html). And two do you have a Ex of using the conversion from a sympy integral to a scipy integral; cause that would be something quite useful

ramvenkat98 commented:
@GProtoZeroW I'm not sure of the answers to those questions - you could ask them on the sympy google group, or make feature requests for them - to my knowledge, Heaviside hasn't been translated from sympy to numpy but it certainly seems useful.
